src/Covenant/ASG.hs

```diff
@@ -951,6 +972,119 @@ cata rAlg rVal =
         t -> throwError . CataNotAnAlgebra $ t
     t -> throwError . CataApplyToNonValT $ t
 
+match ::
+  forall (m :: Type -> Type).
+  (MonadHashCons Id ASGNode m, MonadError CovenantTypeError m, MonadReader ASGEnv m) =>
+  Ref ->
+  Vector Ref ->
+  m Id
+match scrutinee handlers = do
+  scrutNodeTy <- typeRef scrutinee
+  case scrutNodeTy of
+    ValNodeType scrutTy@(Datatype tn args) ->
+      isRecursive scrutTy >>= \case
+        True -> goRecursive tn args
+        False -> goNonRecursive tn args
```

Comment:
This should probably be `if-then-else`. Pattern matching on `Bool` is a bit silly.